modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java

```diff
@@ -1141,178 +1210,340 @@ private void beginSecondCommitPhase(CommitData cd, Stamp commitStamp) throws Exc
       //
       // Its very important the notifications which trigger an observer are deleted after the 2nd
       // phase of commit finishes.
-      getStats().setCommitTs(commitStamp.getTxTimestamp());
-      writeNotificationsAsync(cd, commitStamp.getTxTimestamp());
+      return env.getSharedResources().getOracleClient().getStampAsync().thenApply(commitStamp -> {
+        if (startTs < commitStamp.getGcTimestamp()) {
+          return false;
+        } else {
+          getStats().setCommitTs(commitStamp.getTxTimestamp());
+          return true;
+        }
+      });
     }
+
+    @Override
+    CompletableFuture<Void> getFailureOp(CommitData cd) {
+      return CompletableFuture.supplyAsync(() -> {
```

Comment:
After `throws Exception` is dropped from rollbackLocks, seems like this code should be : 

```java
  CompletableFuture<Void> getFailureOp(CommitData cd) {
      return rollbackLocks(cd);
    }
```